### ISSUE TITLE
EFLOW VM corrupt after Cumulative Updates 

### ISSUE THREAD

**Issue by jbudbo** [Azure/iotedge-eflow#105]
**Describe the Problem**
For the past few weeks, we've noticed that each time a Windows host takes a Cumulate Update, and it seems to be any Cumulative Update, the Eflow VM stops working and cannot be connected to. For example today we have machines that took KB5008218 on 12/15/2021 and the modules running on the Eflow instance are no longer working nor does Eflow. Our only means of resolution is reinstalling Eflow. We've noticed this with each of the most recent versions of IoT Edge including 1.1.2111.17111 however we do still have a machine on 1.0.2.0 which seems to not be effected.

**To Reproduce**
On a stable and running Eflow Host
1. Open an Elevated PowerShell terminal
2. Type `Get-EflowVmEndpoint`
3. Ensure the HealthStatus is in an `OK` state
4. Type `Connect-EflowVm`
5. Ensure you're able to connect
6. Click the Windows button and type `Updates`
7. Click `Check for Updates`
8. Install any Cumulative Update that's available restarting if necessary
9. Repeat steps 1, 2, 3, and 4

**Expected behavior**
The Eflow infrastructure should remain intact across any windows update as advertised
![image](https://user-images.githubusercontent.com/25177311/146413349-634345ac-43c4-4278-94bb-f4b99577bd83.png)

**Actual behavior**
The Eflow infrastructure appears to have entered a corrupt state in which no modules are working and cannot be connected too
![image](https://user-images.githubusercontent.com/25177311/146413706-032bf5ea-42a0-4b77-a003-dedae6cc658c.png)

**Screenshots**
For example this box stopped working today and on the 2nd:
![image](https://user-images.githubusercontent.com/25177311/146414981-368dd801-c6b5-43d8-a894-f452d4d5f151.png)
And each event seems to have a VM Switch being reconstructed each time this happens:
![image](https://user-images.githubusercontent.com/25177311/146415152-6a85373c-5255-4bc2-a4c7-10e55791cbe2.png)

**Windows Host OS (please complete the following information):**
 - Edition: Windows 10 Enterprise LTSC
 - Version: 1809
 - OS build: 17763.2366
 - Azure IoT Edge LTS: 1.1.2111.17111, 1.1.2110.03111
 
**Additional context**
I currently suspect this is an issue that has to do with a change in the VM networking given some of the log entries in the windows event logs as well as the error messages being seen. It appears something is causing changes to the Default Switch being used by Eflow which isn't propagated or something. Though this has been happening for over a month now, the issue is now compounded by the expiration of a code signing cert which causes further pain when attempting to re-deploy eflow:
![image](https://user-images.githubusercontent.com/25177311/146414278-dd115daf-d52c-43a1-a3f7-3caa219dc3dc.png)
Though two certs exist and one is valid, the Deploy-Eflow command doesn't always pick the valid certificate causing the deploy to fail.
![image](https://user-images.githubusercontent.com/25177311/146418759-1dc4989c-158d-4dbc-9cb2-cbdb7eeb247f.png)
I'm able to work around this by manually removing the expired cert

**Comment by jbudbo** [comment]
If it helps, I also have a local VM setup in which I can replicate this 100% of the time

**Comment by jbudbo** [comment]
After reading through some of the issues more, when we enter the bad state, using the command `hcsdiag list` produces no results at all:
![image](https://user-images.githubusercontent.com/25177311/146431416-287e9c44-3ef7-4ac7-8b41-323e98f76708.png)
Whereas prior to the update(s) it works:
![image](https://user-images.githubusercontent.com/25177311/146431661-f6eb8762-1954-4952-a79e-8699c94c11be.png)

**Comment by jbudbo** [comment]
As I continue to dig into this a little more, I believe more heavily that it's an issue stemming from the drop of the VM network endpoint during the update. In my repeatable VM sandbox I am seeing the following prior to update
![image](https://user-images.githubusercontent.com/25177311/146446830-206b9f4c-9384-4dbc-8924-f57c0e834004.png)
![image](https://user-images.githubusercontent.com/25177311/146446909-c67f5586-927d-4c6b-a616-a9cd8117ec4a.png)
![image](https://user-images.githubusercontent.com/25177311/146446947-2f57df79-bd4f-43b4-aa29-510e61c67540.png)
![image](https://user-images.githubusercontent.com/25177311/146446995-4545b847-799c-40dc-94ba-2f398c4ac1eb.png)
![image](https://user-images.githubusercontent.com/25177311/146447010-9cc27bf9-9904-4db2-b068-6527053c42fd.png)
However after installing KB5008215 for example, none of the `hcsdiag` commands work and I'm seeing the following
![image](https://user-images.githubusercontent.com/25177311/146447150-8b50400c-e820-490f-8613-cca7d0f6fcb0.png)
Notice that `Get-HnsEndpoint` no longer returns anything and for the data returned from `Get-HnsNetwork` notice that now
* CurrentEndpointCount = 0
* NatName has changed
* TotalEndpoints = 0
* Subnets have changed from 172.23.0.0/20 to 172.27.224.0/20

**Comment by fcabrera23** [comment]
Hi @jbudbo,

Thanks for all the detailed information. We will try to reproduce the issue in our environment and reach back. In the meantime, could you please answer the following questions:

1) Are you using a WiFi-connected device? Are you changing between different WiFi networks?
2) After the update, can you try doing a `Stop-EflowVm` and `Start`EflowVm` and see if the problem persists?

Thanks,
Francisco

**Comment by jbudbo** [comment]
Hi @fcabrera23, these devices are not connected via Wifi and neither of those commands work without error. I'll work on getting a screenshot but essentially they report that the VM can't be reached.

**Comment by ngmartins** [comment]
Hello, @jbudbo and @fcabrera23 we are also experiencing this issue in our staging environments for a week now.
We opened a ticker to try to realize the root cause. 
Until now we didn't associate it with windows updates. I will try to check this relationship on our side. 

@jbudbo how do you revert this status? We just manage to do it by uninstalling the .msi and re-deploying the VM.

Thanks, 
Nuno

**Comment by jbudbo** [comment]
@ngmartins it took us a while to pin it down to updates. And oddly, from what I can tell not just any update but Cumulative updates for one reason or another. When a system gets in this state the only thing we've found we can do is uninstall and rebuilding like you've mentioned you're doing. I've tried to find a way to maybe just rebuild the virtual network adapter but haven't had any luck (or I'm just not doing something right).

**Comment by ngmartins** [comment]
Thanks once again for your help.

About the updates: Our devices' updates are managed by update rings. Just checked the updates' history and there are no updates on this device since June. But this device got stuck in that state 4 times since last Friday.
![image](https://user-images.githubusercontent.com/24669761/146546423-89f73606-3740-46c2-8c0c-5529cf412361.png)

I manually forced the update of today and after de updates, it is still possible to connect to the VM (didn't course the faulty status for us).

**Comment by jbudbo** [comment]
@ngmartins, that's really interesting. In your system event logs do you see records indicating that a VM Network Switch we removed and re-added like in my screenshot above?

**Comment by ngmartins** [comment]
For the today's updates, I can see Creation, Initialization, and Connection events from a Hyper-V-VmSwitch source. However, I cannot see Removing Events.
 (EventIds - 102; 232; 233; 264 and 7)

**Comment by fcabrera23** [comment]
Hi @ngmartins , @jbudbo

Can you please try using External Switch and see if the problem persists?

Thanks,
Francisco

**Comment by jbudbo** [comment]
@fcabrera23, so I tried this in my sandbox device and it took me a little while to figure out the steps as I haven't done a deployment on an external network before.

Here is what I did:
1. Open Hyper-V Manager and open Virtual Switch Manager
2. Create a new Virtual Switch named `nvw` and set it an External switch bound to my NIC
3. Uninstall Eflow
4. Reinstall Eflow and redeploy using `Deploy-Eflow -acceptEula yes -acceptOptionalTelemetry no -vswitchName nvw -vswitchType External`
    * At this point the Deployment partially succeeded as my sandbox doesn't seem to be providing an IP to the new switch but the new Eflow VM does show as on and I can use `hcsdiag console` to gain access to it
5. Install my pending Cumulative Update (KB5008215) and reboot

Unlike what I see in the issue I've raised here, after the reboot my Eflow VM seems to be in essentially the same state as I saw in Step 4. Meaning that after reboot my Eflow device is once again in a started state (unlike what I've reported initially) and I can gain access to it using `hcsdiag` (unlike what I've reported initially) so I'm assuming if I were able to figure out my sandbox networking issue then this would have worked. That said however, I do still see entries in my event log indicating that both my `Default Switch` and `nvw` virtual switches were removed and reconstructed during the update. It's just that now my Eflow VM isn't drawing connectivity from the internal system rather the external network. I do find it strange though that in my first report I can't even list the device using `hcsdiag` but I can with this workflow.

**Comment by jbudbo** [comment]
Good morning @fcabrera23, any other thoughts on this one? I suspect your suggestion to move to an external switch may work however I don't know that this is the topology we'd want to use in all situations.

**Comment by fcabrera23** [comment]
Hi @jbudbo,

Were you able to confirm if the problem persists when using External Switch during this long period?  We are working with the Hyper-V team to repro the issue, as we think it's related to the Default Switch.

Thanks,
Francisco

**Comment by jbudbo** [comment]
@fcabrera23, sorry it's taken so long to get back to you. I'm still looking over our systems to find one that has taken a CU during the break and see if it's still up. That said, we haven't had any machine go down during the long period of time since we last connected (except one but it seems unrelated and with different symptoms). I need to try and setup my internal dev box and see if it sustains itself.

**Comment by fcabrera23** [comment]
Hi @jbudbo,

I'm glad to hear that the update worked and everything is working with External Switch. We've found an issue with the Default Switch and we plan to address it in the upcoming updates. In the meantime, I suggest keep using the External Switch for the best support and stability.

Thanks,
Francisco

**Comment by jbudbo** [comment]
@fcabrera23, that's excellent news! We can probably keep things moving forward for the short term with an External switch and attempting to stop Hyper-V during update (if we have the opportunity) but any thoughts on what timeframe we'd be looking at for the issue you've identified?

**Comment by fcabrera23** [comment]
@jbudbo,

There's no need for stopping Hyper-V and the EFLOW VM for the update. It should work without any action. 
Regarding the timeframe, we are aiming to include the fix in our next update, end of Jan, although we haven't confirmed it yet.

Thanks,
Francisco

**Comment by jbudbo** [comment]
hey @fcabrera23,
We had another client go down today even though we stopped updates temporarily on their box. Same symptoms in that I can't get connected to the Eflow VM. Even if I reboot the machine and/or stop and restart the WSSDAgent nothing seems to work. It's as if the VM is "gone". When I run `hcsdiag list` I get nothing back. I managed to dump the result of a call to `Get-EflowVM` to JSON.
In doing so I get the following exceptions:
 * Could not resolve virtual machine IP address at line 3881                    
 * Cannot index into a null array. at line 1558
 And the resulting Json looks as follows (I've expanded the nested error message a little)
```json
{
	"VmConfiguration": {
		"ID": "14d0b3530c16ee9",
		"name": "SB-7080-001-EFLOW",
		"properties": {
			"HighAvailabilityState": "UNKNOWN_HA_STATE",
			"disableHighAvailability": false,
			"hardwareProfile": {
				"customsize": {
					"cpucount": 2,
					"gpucount": 0,
					"memorymb": 4096
				},
				"vmSize": "Custom"
			},
			"isPlaceholder": false,
			"networkProfile": {
				"networkInterfaces": [
					{
						"virtualNetworkInterfaceReference": "SB-7080-001-EFLOWInterface"
					}
				]
			},
			"osProfile": {
				"computerName": "SB-7080-001-EFLOW",
				"linuxConfiguration": {
					"disablePasswordAuthentication": true
				},
				"osbootstrapengine": "CloudInit",
				"windowsConfiguration": {
					"rdp": {
						"DisableRDP": null
					}
				}
			},
			"provisioningState": "UPDATED",
			"securityProfile": {
				"enableTPM": false
			},
			"statuses": {
				"Error": "Message:\"Not Found
					Virtual Machine [SB-7080-001-EFLOW] not found
					github.com/microsoft/moc/pkg/errors.Wrapf
						/home/johnson/go/pkg/mod/github.com/johnsonshih/moc@v0.10.8-alpha.10.0.20210526222753-7f9ed43aa602/pkg/errors/errors.go:118
					github.com/microsoft/wssdagent/pkg/nodeagent/services/compute/virtualmachine/hcs.(*Client).StopVirtualMachine
						/home/johnson/repos/wssdagent.js/pkg/nodeagent/services/compute/virtualmachine/hcs/hcs.go:370
					github.com/microsoft/wssdagent/pkg/nodeagent/services/compute/virtualmachine.(*Client).Stop
						/home/johnson/repos/wssdagent.js/pkg/nodeagent/services/compute/virtualmachine/client.go:556
					github.com/microsoft/wssdagent/pkg/nodeagent/services/compute/virtualmachine.(*VirtualMachineProvider).Stop
						/home/johnson/repos/wssdagent.js/pkg/nodeagent/services/compute/virtualmachine/virtualmachine.go:89
					github.com/microsoft/wssdagent/pkg/nodeagent/server.(*virtualmachineAgentServer).Operate
						/home/johnson/repos/wssdagent.js/pkg/nodeagent/server/virtualmachine.go:95
					github.com/microsoft/moc/rpc/nodeagent/compute._VirtualMachineAgent_Operate_Handler
						/home/johnson/go/pkg/mod/github.com/johnsonshih/moc@v0.10.8-alpha.10.0.20210526222753-7f9ed43aa602/rpc/nodeagent/compute/moc_nodeagent_virtualmachine.pb.go:1177
					google.golang.org/grpc.(*Server).processUnaryRPC
						/home/johnson/go/pkg/mod/google.golang.org/grpc@v1.26.0/server.go:1024
					google.golang.org/grpc.(*Server).handleStream
						/home/johnson/go/pkg/mod/google.golang.org/grpc@v1.26.0/server.go:1313
					google.golang.org/grpc.(*Server).serveStreams.func1.1
						/home/johnson/go/pkg/mod/google.golang.org/grpc@v1.26.0/server.go:722
					runtime.goexit
						/usr/lib/go-1.15/src/runtime/asm_amd64.s:1374\" ",
				"HealthState": "",
				"PowerState": "Running",
				"ProvisionState": "currentState:UPDATED previousState:UPDATING ",
				"Version": "1642096472806222000"
			},
			"storageProfile": {
				"dataDisks": [],
				"osDisk": {
					"vhd": "AzureIoTEdgeForLinux-v1-EFLOW"
				},
				"vmConfigContainerName": ""
			}
		},
		"tags": {
			"SB-7080-001-EFLOW-id": "dcff55a7-7499-11ec-b129-a4bb6dd3c68f"
		}
	},
	"VmPowerState": "Running",
	"EdgeRuntimeVersion": {},
	"EdgeRuntimeStatus": {},
	"SystemStatistics": {}
}
```

Do you think this is any way related to what I've brought up in this chain?

**Comment by fcabrera23** [comment]
Hi @jbudbo,

- Did the VM stop without reboot or update? 
- Is this deployment also using Default Switch?
- Do you have access to this VM so that we can do some further troubleshooting?

Thanks,
Francisco

**Comment by jbudbo** [comment]
Hi @fcabrera23,
* It does appear to have stopped on it's own accord. Updates have been paused and the machine had been up for some time (I've rebooted it in an attempt to resolve the issues)
* It is using the Default (and internal) Switch
* I do have access to the host windows machine to troubleshoot anything you need

**Comment by fcabrera23** [comment]
Hi @jbudbo,

Thanks for the information provided. From the events you shared, there was a hard shutdown (due to lack of power) of the host OS that causes the same Default Switch issue that Cumulative updates. This issue will be fixed in our next update.

Thanks,
Francisco

**Comment by jbudbo** [comment]
Hey @fcabrera23,
Not trying to be impatient or anything but I just wanted to check in on the patch/update y'all were thinking may be available by the end of Jan. Updates are set to resume across our fleet of devices in the next week or so.
We've been relatively stable since pausing them but we'd like to avoid getting too far behind.
Thanks for everything!
~Justin

**Comment by jagadishmurugan** [comment]
Hi @jbudbo,
That fix should be available in Jan release - in 1st week of Feb/2022.

Regards,
Jagadish

**Comment by fcabrera23** [comment]
Hi @jbudbo ,

This issue was fixed with the [EFLOW 1.1.2201.04022 update](https://aka.ms/AzEFLOW-1_1_2201_04022).

Thanks,
Francisco
